File: zero-server/src/main/kotlin/com/zeroday/service/DailyLoginService.kt
package com.zeroday.service

import com.zeroday.model.NotificationType
import com.zeroday.model.Player

/**
 * Tracks daily login streaks, grants streak rewards, calculates offline
 * resource accumulation, and auto-rotates challenges on login.
 *
 * All methods are thread-safe because they operate on [Player] only through
 * [PlayerService.withPlayerAction], which takes the per-player lock.
 */
class DailyLoginService(
    private val playerService: PlayerService,
    private val notificationService: NotificationService? = null,
    private val challengeService: ChallengeService? = null
) {
    companion object {
        /** Maximum offline accumulation window (millis). */
        const val MAX_OFFLINE_MILLIS: Long = 28_800_000L   // 8h

        /** Minimum offline duration to grant idle rewards. */
        const val MIN_OFFLINE_MILLIS: Long = 3_600_000L    // 1h

        /** Resources granted per hour of offline time. */
        const val OFFLINE_CPU_PER_HOUR: Int = 2
        const val OFFLINE_RAM_PER_HOUR: Int = 4
        const val OFFLINE_CREDITS_PER_HOUR_PER_LEVEL: Long = 15

        /** Minimum time between logins to count as a new day (16h). */
        const val STREAK_WINDOW_MILLIS: Long = 57_600_000L

        /** One day in millis (~40h max gap to allow skipping a day). */
        const val DAY_MILLIS: Long = 86_400_000L

        fun dailyStreakReward(streakDay: Int): StreakReward {
            // Cycle every 7 days for predictability.
            val day = ((streakDay - 1) % 7) + 1
            return when (day) {
                1 -> StreakReward(credits = 100, xp = 50)
                2 -> StreakReward(credits = 200, xp = 100)
                3 -> StreakReward(credits = 350, xp = 175)
                4 -> StreakReward(credits = 500, xp = 250)
                5 -> StreakReward(credits = 750, xp = 400)
                6 -> StreakReward(credits = 1_000, xp = 600)
                7 -> StreakReward(credits = 2_000, xp = 1_000, reputation = 5)
                else -> StreakReward(credits = 100, xp = 50)
            }
        }
    }

    data class StreakReward(
        val credits: Long = 0,
        val xp: Long = 0,
        val reputation: Int = 0
    )

    data class LoginResult(
        val streak: Int,
        val isNewStreak: Boolean,
        val streakReward: StreakReward?,
        val offlineReward: OfflineAccumulation?,
        val message: String
    )

    data class OfflineAccumulation(
        val cpuGained: Int,
        val ramGained: Int,
        val creditsGained: Long,
        val offlineHours: Double
    )

    /**
     * Process a player login. Updates streak, grants rewards, and pushes
     * notifications. Must be called after [PlayerService.login] succeeds.
     * Returns null if the player does not exist.
     */
    suspend fun processLogin(
        playerId: String,
        ipAddress: String? = null
    ): LoginResult? {
        return playerService.withPlayerAction(playerId) { player ->
            val now = System.currentTimeMillis()
            val offline = calculateOffline(player, now)
            val streakResult = updateStreak(player, now)
            val reward = streakResult.reward

            // Apply offline accumulation.
            if (offline != null) {
                player.cpu = (player.cpu + offline.cpuGained).coerceAtMost(player.maxCpu)
                player.ram = (player.ram + offline.ramGained).coerceAtMost(player.maxRam)
                player.credits += offline.creditsGained
                player.lifetimeCreditsEarned += offline.creditsGained
            }

            // Apply streak reward.
            if (reward != null) {
                player.credits += reward.credits
                player.lifetimeCreditsEarned += reward.credits
                player.experience += reward.xp
                player.reputation += reward.reputation
                checkForLevelUp(player)
            }

            // Update login tracking.
            player.lastLoginAt = now
            player.totalLogins++
            if (streakResult.isNewStreak) player.loginStreak = streakResult.streak
            if (player.loginStreak > player.longestStreak) player.longestStreak = player.loginStreak
            player.lastLoginIp = ipAddress ?: player.lastLoginIp

            // Auto-rotate challenges.
            challengeService?.rotateIfNeeded(player)

            // Build message.
            val parts = mutableListOf<String>()
            if (offline != null && offline.creditsGained > 0) {
                parts.add("While you were away (+${offline.offlineHours.format(1)}h): +${offline.creditsGained}cr, +${offline.cpuGained}CPU, +${offline.ramGained}RAM")
            }
            if (reward != null) {
                parts.add("Day ${streakResult.streak} streak reward: +${reward.credits}cr, +${reward.xp}xp${if (reward.reputation > 0) ", +${reward.reputation} rep" else ""}")
                parts.add("Next reward in: day ${(streakResult.streak % 7) + 1}")
            }
            val message = if (parts.isEmpty()) "Welcome back!" else parts.joinToString("\n")

            // Push notifications.
            if (reward != null) {
                notificationService?.push(
                    player = player,
                    type = NotificationType.DAILY_LOGIN,
                    title = "Daily Login Streak: Day ${streakResult.streak}",
                    message = "You earned +${reward.credits}cr, +${reward.xp}xp${if (reward.reputation > 0) " and +${reward.reputation} rep" else ""} for logging in today!",
                    data = mapOf(
                        "streak" to streakResult.streak.toString(),
                        "credits" to reward.credits.toString(),
                        "xp" to reward.xp.toString(),
                        "reputation" to reward.reputation.toString()
                    )
                )
            }

            LoginResult(
                streak = streakResult.streak,
                isNewStreak = streakResult.isNewStreak,
                streakReward = reward,
                offlineReward = offline,
                message = message
            )
        } ?: run {
            // Player was deleted between login and here.
            LoginResult(
                streak = 0,
                isNewStreak = false,
                streakReward = null,
                offlineReward = null,
                message = "Player not found"
            )
        }
    }

    /**
     * Non-mutating: calculate what the player would have earned while
     * offline without applying anything.
     */
    fun calculateOffline(player: Player, now: Long = System.currentTimeMillis()): OfflineAccumulation? {
        if (player.lastLoginAt <= 0L) return null
        val elapsed = now - player.lastLoginAt
        if (elapsed < MIN_OFFLINE_MILLIS) return null
        val capped = elapsed.coerceAtMost(MAX_OFFLINE_MILLIS)
        val hours = capped.toDouble() / 3_600_000.0
        return OfflineAccumulation(
            cpuGained = (hours * OFFLINE_CPU_PER_HOUR).toInt().coerceAtLeast(0),
            ramGained = (hours * OFFLINE_RAM_PER_HOUR).toInt().coerceAtLeast(0),
            creditsGained = (hours * player.level * OFFLINE_CREDITS_PER_HOUR_PER_LEVEL).toLong().coerceAtLeast(0),
            offlineHours = hours
        )
    }

    // ---- internal ----

    private data class StreakResult(
        val streak: Int,
        val isNewStreak: Boolean,
        val reward: StreakReward?
    )

    /**
     * Evaluate and return the new streak value and reward. Does NOT mutate
     * [player] — the caller [processLogin] applies rewards.
     */
    private fun updateStreak(player: Player, now: Long): StreakResult {
        val sinceLast = now - player.lastLoginAt

        // First ever login.
        if (player.lastLoginAt <= 0L) {
            return StreakResult(1, true, dailyStreakReward(1))
        }

        // Same calendar day: no streak change, no reward.
        if (sinceLast < STREAK_WINDOW_MILLIS) {
            return StreakResult(player.loginStreak.coerceAtLeast(1), false, null)
        }

        // Consecutive day (within ~40h window to allow some flexibility):
        // the player logged in yesterday or the day before.
        val maxGap = STREAK_WINDOW_MILLIS + DAY_MILLIS // ~40h
        if (sinceLast < maxGap) {
            val newStreak = player.loginStreak + 1
            return StreakResult(newStreak, true, dailyStreakReward(newStreak))
        }

        // Streak broken: more than 40h since last login.
        return StreakResult(1, true, dailyStreakReward(1))
    }

    private fun checkForLevelUp(player: Player) {
        var leveledUp = false
        var reachedLevel = player.level
        while (player.experience >= player.experienceToNext) {
            player.experience -= player.experienceToNext
            player.level++
            player.experienceToNext = (player.experienceToNext * 1.5).toLong()
            player.maxCpu += 5
            player.maxRam += 16
            player.maxBandwidth += 2
            player.cpu = player.maxCpu
            player.ram = player.maxRam
            player.bandwidth = player.maxBandwidth
            player.skillPoints += if (player.level % 5 == 0) 2 else 1
            leveledUp = true
            reachedLevel = player.level
        }
        if (leveledUp) player.lastLevelNotified = reachedLevel
    }
}

/** Format double to N decimal places without pulling in extra deps. */
private fun Double.format(decimals: Int): String = "%.${decimals}f".format(this)

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/DailyLoginService.cs
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Services;

public class DailyLoginService
{
    private readonly PlayerService _playerService;
    private readonly NotificationService? _notificationService;
    private readonly ChallengeService? _challengeService;

    public const long MaxOfflineMillis = 28_800_000L;
    public const long MinOfflineMillis = 3_600_000L;
    public const int OfflineCpuPerHour = 2;
    public const int OfflineRamPerHour = 4;
    public const long OfflineCreditsPerHourPerLevel = 15;
    public const long StreakWindowMillis = 57_600_000L;
    public const long DayMillis = 86_400_000L;

    public DailyLoginService(
        PlayerService playerService,
        NotificationService? notificationService = null,
        ChallengeService? challengeService = null)
    {
        _playerService = playerService;
        _notificationService = notificationService;
        _challengeService = challengeService;
    }

    public static StreakReward DailyStreakReward(int streakDay)
    {
        var day = ((streakDay - 1) % 7) + 1;
        return day switch
        {
            1 => new StreakReward(100, 50),
            2 => new StreakReward(200, 100),
            3 => new StreakReward(350, 175),
            4 => new StreakReward(500, 250),
            5 => new StreakReward(750, 400),
            6 => new StreakReward(1000, 600),
            7 => new StreakReward(2000, 1000, 5),
            _ => new StreakReward(100, 50)
        };
    }

    public async Task<LoginResult?> ProcessLogin(string playerId, string? ipAddress = null)
    {
        var result = await _playerService.WithPlayerAction<LoginResult>(playerId, player =>
        {
            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var offline = CalculateOffline(player, now);
            var streakResult = UpdateStreak(player, now);
            var reward = streakResult.Reward;

            if (offline != null)
            {
                player.Cpu = Math.Min(player.Cpu + offline.CpuGained, player.MaxCpu);
                player.Ram = Math.Min(player.Ram + offline.RamGained, player.MaxRam);
                player.Credits += offline.CreditsGained;
                player.LifetimeCreditsEarned += offline.CreditsGained;
            }

            if (reward != null)
            {
                player.Credits += reward.Credits;
                player.LifetimeCreditsEarned += reward.Credits;
                player.Experience += reward.Xp;
                player.Reputation += reward.Reputation;
                CheckForLevelUp(player);
            }

            player.LastLoginAt = now;
            player.TotalLogins++;
            if (streakResult.IsNewStreak) player.LoginStreak = streakResult.Streak;
            if (player.LoginStreak > player.LongestStreak) player.LongestStreak = player.LoginStreak;
            player.LastLoginIp = ipAddress ?? player.LastLoginIp;

            _challengeService?.RotateIfNeeded(player);

            var parts = new List<string>();
            if (offline != null && offline.CreditsGained > 0)
            {
                parts.Add($"While you were away (+{offline.OfflineHours:F1}h): +{offline.CreditsGained}cr, +{offline.CpuGained}CPU, +{offline.RamGained}RAM");
            }
            if (reward != null)
            {
                var repStr = reward.Reputation > 0 ? $", +{reward.Reputation} rep" : "";
                parts.Add($"Day {streakResult.Streak} streak reward: +{reward.Credits}cr, +{reward.Xp}xp{repStr}");
                parts.Add($"Next reward in: day {(streakResult.Streak % 7) + 1}");
            }

            var message = parts.Count > 0 ? string.Join("\n", parts) : "Welcome back!";

            if (reward != null)
            {
                _notificationService?.AddNotification(player.Id, new Notification
                {
                    Type = NotificationType.DAILY_LOGIN,
                    Title = $"Daily Login Streak: Day {streakResult.Streak}",
                    Message = $"You earned +{reward.Credits}cr, +{reward.Xp}xp{(reward.Reputation > 0 ? $" and +{reward.Reputation} rep" : "")} for logging in today!",
                    Data = new Dictionary<string, string>
                    {
                        ["streak"] = streakResult.Streak.ToString(),
                        ["credits"] = reward.Credits.ToString(),
                        ["xp"] = reward.Xp.ToString(),
                        ["reputation"] = reward.Reputation.ToString()
                    }
                });
            }

            return new LoginResult(
                streakResult.Streak,
                streakResult.IsNewStreak,
                reward,
                offline,
                message
            );
        });

        return result ?? new LoginResult(0, false, null, null, "Player not found");
    }

    public OfflineAccumulation? CalculateOffline(Player player, long? nowMs = null)
    {
        var now = nowMs ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (player.LastLoginAt <= 0L) return null;

        var elapsed = now - player.LastLoginAt;
        if (elapsed < MinOfflineMillis) return null;

        var capped = Math.Min(elapsed, MaxOfflineMillis);
        var hours = capped / 3_600_000.0;

        return new OfflineAccumulation(
            Math.Max(0, (int)(hours * OfflineCpuPerHour)),
            Math.Max(0, (int)(hours * OfflineRamPerHour)),
            Math.Max(0, (long)(hours * player.Level * OfflineCreditsPerHourPerLevel)),
            hours
        );
    }

    private StreakResult UpdateStreak(Player player, long now)
    {
        var sinceLast = now - player.LastLoginAt;

        if (player.LastLoginAt <= 0L)
            return new StreakResult(1, true, DailyStreakReward(1));

        if (sinceLast < StreakWindowMillis)
            return new StreakResult(Math.Max(1, player.LoginStreak), false, null);

        var maxGap = StreakWindowMillis + DayMillis;
        if (sinceLast < maxGap)
        {
            var newStreak = player.LoginStreak + 1;
            return new StreakResult(newStreak, true, DailyStreakReward(newStreak));
        }

        return new StreakResult(1, true, DailyStreakReward(1));
    }

    private static void CheckForLevelUp(Player player)
    {
        var leveledUp = false;
        var reachedLevel = player.Level;

        while (player.Experience >= player.ExperienceToNext)
        {
            player.Experience -= player.ExperienceToNext;
            player.Level++;
            player.ExperienceToNext = (long)(player.ExperienceToNext * 1.5);
            player.MaxCpu += 5;
            player.MaxRam += 16;
            player.MaxBandwidth += 2;
            player.Cpu = player.MaxCpu;
            player.Ram = player.MaxRam;
            player.Bandwidth = player.MaxBandwidth;
            player.SkillPoints += player.Level % 5 == 0 ? 2 : 1;
            leveledUp = true;
            reachedLevel = player.Level;
        }

        if (leveledUp) player.LastLevelNotified = reachedLevel;
    }

    public record StreakReward(long Credits = 0, long Xp = 0, int Reputation = 0);

    public record LoginResult(
        int Streak,
        bool IsNewStreak,
        StreakReward? StreakReward,
        OfflineAccumulation? OfflineReward,
        string Message
    );

    public record OfflineAccumulation(
        int CpuGained,
        int RamGained,
        long CreditsGained,
        double OfflineHours
    );

    private record StreakResult(int Streak, bool IsNewStreak, StreakReward? Reward);
}
